pkg/controller/container-runtime-config/helpers_test.go

```diff
@@ -39,98 +39,113 @@ func TestScopeMatchesRegistry(t *testing.T) {
+					{Source: "source.example.com", Mirrors: []string{"z1.example.com", "y2.example.com"}},
 				},
 				{
-					{"y2.example.net", "x3.example.net"},
-					{"y2.example.com", "x3.example.com"},
+					{Source: "source.example.net", Mirrors: []string{"y2.example.net", "x3.example.net"}},
+					{Source: "source.example.com", Mirrors: []string{"y2.example.com", "x3.example.com"}},
+				},
+			},
+			result: []apioperatorsv1alpha1.RepositoryDigestMirrors{
+				{Source: "source.example.com", Mirrors: []string{"z1.example.com", "y2.example.com", "x3.example.com"}},
+				{Source: "source.example.net", Mirrors: []string{"z1.example.net", "y2.example.net", "x3.example.net"}},
+			},
+		},
+		{
+			// This is not technically impossible, and it could be in principle used to set up last-fallback mirrors that
+			// are only accessed if the source is not available.
+			// WARNING: The order in this case is unspecified by the ICSP specification, and may change at any time;
+			// this test case only ensures that the corner case is handled reasonably, and that the output is stable
+			// (i.e. the operator does not cause unnecessary changes in output objects.)
+			name: "Source included in mirrors",
+			input: [][]apioperatorsv1alpha1.RepositoryDigestMirrors{
+				{
+					{Source: "source.example.com", Mirrors: []string{"z1.example.com", "source.example.com", "y2.example.com"}},
+					{Source: "source.example.com", Mirrors: []string{"source.example.com", "y2.example.com", "x3.example.com"}},
 				},
 			},
-			result: [][]string{
-				{"z1.example.com", "y2.example.com", "x3.example.com"},
-				{"z1.example.net", "y2.example.net", "x3.example.net"},
+			result: []apioperatorsv1alpha1.RepositoryDigestMirrors{
+				{Source: "source.example.com", Mirrors: []string{"z1.example.com", "source.example.com", "y2.example.com", "x3.example.com"}},
 			},
 		},
 		// More complex mirror set combinations are mostly tested in TestTopoGraph
 		{
 			name: "Example",
-			input: [][][]string{
+			input: [][]apioperatorsv1alpha1.RepositoryDigestMirrors{
 				{ // Vendor-provided default configuration
-					{"registry1.vendor.com", "registry2.vendor.com"},
+					{Source: "source.vendor.com", Mirrors: []string{"registry2.vendor.com"}},
 				},
 				{ // Vendor2-provided default configuration
-					{"registry1.vendor2.com", "registry2.vendor2.com"},
+					{Source: "source.vendor2.com", Mirrors: []string{"registry1.vendor2.com", "registry2.vendor2.com"}},
 				},
 				{ // Admin-configured local mirrors:
-					{"local-mirror.example.com", "registry1.vendor.com"},
-					{"local-mirror2.example.com", "registry2.vendor2.com", "registry1.vendor2.com"}, // Opposite order of the vendor’s mirrors
+					{Source: "source.vendor.com", Mirrors: []string{"local-mirror.example.com"}},
+					// Opposite order of the vendor’s mirrors.
+					// WARNING: The order in this case is unspecified by the ICSP specification, and may change at any time;
```

Comment:
This comment is identical to the previous test comment. Are these 2 tests testing the same thing? Trying to follow along.

Reply:
They test two different things:
- `Source` is included in `Mirrors`
- Two `RepositoryDigestMirrors` items contain items in `Mirrors` in conflicting order.

ICSP does not _document_ how such cases are resolved, it’s not part of the API guarantee.

OTOH, MCC does have an interest of keeping the output stable over time, so that the controller does not issue unnecessary updates to the configuration.  So, the implementation sorts the items to behave consistently over time, and these tests hard-code that implementation detail.

The warnings are added to decrease the chance that someone will read the code and the tests, and conclude that the current order is an API promise. Sure, it’s pretty unlikely that anyone will read that, if you think the warnings are too confusing to be worth that, I’d be happy to remove them.

Reply:
I think the warnings are a good idea, since like you said there is room for confusion. Just wanted to make sure that the warning did need to be present in two places.